### DISCUSSION TITLE
If links are prefetched, why does it take so long to open a page?

### DISCUSSION THREAD

**Discussion by YCMitch** [vercel/next.js#19583] [Help] [answered]
My understanding of NextJS is that it prefetches pages that are within the viewport, but in our case it still takes 3-4 seconds to load the subsequent page (even when deployed to production). 

Is there anything we can do to speed this up? We're used to Gatsby where the load time is somewhere near zero.

We are using `getInitialProps` to make server-side queries, but I'd presume if it's prefetched then that shouldn't slow it down, right?

**Comment by jamesmosier** [comment, marked as answer]
[Prefetching](https://web.dev/link-prefetch/) just prefetches the assets (JavaScript, CSS). It does not execute the code on the client nor on the server.

Gatsby is near zero because the pages are statically generated, which Next does (and then some!) via [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation). 

To decrease page load times, use `getStaticProps` to generate your page at build time or reduce the amount of logic you have in your `getInitialProps` and instead fetch data on the client after loading some of the information on the page (so the user can see *something* while your fetch data).

**Reply by YCMitch** [reply]
Oh dang okay. So there's no way to prefetch the queries too? 

We're steering away from SSG as clients always complain about having to wait for recompilation, but I dearly miss the crazy-quick load times. We still kinda need most queries to run server-side though (as it all needs to be crawlable). 

Is it perhaps feasible to check server-side whether a query has already been run? A lot of them are repeated across pages (menu items etc), which probably only need to be run once.

**Reply by jamesmosier** [reply]
There is no way to prefetch the the queries as well. You'd want to implement some caching strategies outside of Next to ensure your pages are only fetched and the queries are run when the cache is stale

**Reply by leerob** [reply]
Yep, as @jamesmosier said here, you're comparing apples to oranges (SSG Gatsby vs. Non-SSG Next). 

> We're steering away from SSG as clients always complain about having to wait for recompilation, but I dearly miss the crazy-quick load times. We still kinda need most queries to run server-side though (as it all needs to be crawlable).

![image](https://user-images.githubusercontent.com/9113740/100694780-58346e80-3355-11eb-99a3-5af996923cbd.png)

You can use SSG with getStaticPaths/Props and have fast load times as well as fast builds. These are all pre-rendered on the server (because of SSG). You can specify a certain set of pages to be statically generated at build time (in getStaticPaths/Props) and then use `fallback` to handle others at request time (keeping builds fast).
